crux-core/src/crux/kv/tx_log.clj

```diff
@@ -30,42 +30,52 @@
 (defn decode-tx-event-key-from [^DirectBuffer k]
   (assert (= (+ c/index-id-size Long/BYTES Long/BYTES) (.capacity k)) (mem/buffer->hex k))
   (assert (tx-event-key? k))
-  {:crux.tx/tx-id (.getLong k c/index-id-size ByteOrder/BIG_ENDIAN)
-   :crux.tx/tx-time (c/reverse-time-ms->date (.getLong k (+ c/index-id-size Long/BYTES) ByteOrder/BIG_ENDIAN))})
-
-(defn- ingest-tx [tx-ingester tx tx-events]
-  (let [in-flight-tx (db/begin-tx tx-ingester tx nil)]
-    (if (db/index-tx-events in-flight-tx tx-events)
-      (db/commit in-flight-tx)
-      (db/abort in-flight-tx))))
-
-(defn- submit-tx [tx-events
-                  {:keys [^ExecutorService tx-submit-executor
-                          ^ExecutorService tx-ingest-executor
-                          kv-store tx-ingester fsync?]}]
+  {::tx/tx-id (.getLong k c/index-id-size ByteOrder/BIG_ENDIAN)
+   ::tx/tx-time (c/reverse-time-ms->date (.getLong k (+ c/index-id-size Long/BYTES) ByteOrder/BIG_ENDIAN))})
+
+(defn- latest-submitted-tx [kv-store]
+  (when-let [tx-id (kvi/read-meta kv-store :crux.kv-tx-log/latest-submitted-tx-id)]
+    {::tx/tx-id tx-id}))
+
+(defn- submit-tx [tx-events {:keys [^ExecutorService tx-submit-executor kv-store fsync? subscriber-handler]}]
   (if (.isShutdown tx-submit-executor)
     ::closed
 
     ;; this needs to remain `:crux.kv-tx-log/latest-submitted-tx-id` because we're a TxLog
-    (let [tx-time (Date.)
-          tx-id (inc (or (kvi/read-meta kv-store :crux.kv-tx-log/latest-submitted-tx-id) -1))
-          next-tx {:crux.tx/tx-id tx-id, :crux.tx/tx-time tx-time}]
-      (kv/store kv-store [[(encode-tx-event-key-to nil next-tx)
-                           (mem/->nippy-buffer tx-events)]
-                          (kvi/meta-kv :crux.kv-tx-log/latest-submitted-tx-id tx-id)])
+    (try
+      (let [tx-time (Date.)
+            tx-id (inc (or (kvi/read-meta kv-store :crux.kv-tx-log/latest-submitted-tx-id) -1))
+            next-tx {::tx/tx-id tx-id, ::tx/tx-time tx-time}]
+        (kv/store kv-store [[(encode-tx-event-key-to nil next-tx)
+                             (mem/->nippy-buffer tx-events)]
+                            (kvi/meta-kv :crux.kv-tx-log/latest-submitted-tx-id tx-id)])
 
-      (when (and tx-ingest-executor tx-ingester)
-        (.submit tx-ingest-executor
-                 ^Runnable #(ingest-tx tx-ingester next-tx tx-events)))
+        (when fsync?
+          (kv/fsync kv-store))
 
-      (when fsync?
-        (kv/fsync kv-store))
+        (tx-sub/notify-tx! subscriber-handler next-tx)
 
-      next-tx)))
+        next-tx)
+      (catch Exception e
+        (.printStackTrace e)
```

Comment:
Presumably here for debugging - can take out?

Reply:
cheers @jonpither :pray: